### ISSUE TITLE
Bug Report: Optimizer not set when backend is changed

### ISSUE THREAD

**Issue by matthewfeickert** [diana-hep/pyhf#99]
# Description

If using a backend other than `numpy_backend` currently we have to manually set the optimizer. However, this should be done automatically when the backend is changed.

Otherwise this causes an error, as shown below with `pyhf.runOnePoint()`.

# Expected Behavior

When changing `pyhf.tensorlib` have this change be automatically detected and also change `pyhf.optimizer` appropriately.

# Actual Behavior

When changing `pyhf.tensorlib` `pyhf.optimizer` must be manually set. If it is not, then errors can be caused, as seen below.

# Steps to Reproduce

```python
import pyhf
from pyhf.tensor.tensorflow_backend import tensorflow_backend
from pyhf.simplemodels import hepdata_like
import tensorflow as tf

if __name__ == '__main__':
    default_backend = pyhf.tensorlib
    pyhf.tensorlib = tensorflow_backend(session=tf.Session())

    n_bins = 5
    binning = [n_bins, -0.5, n_bins + 0.5]
    data = [120.0] * n_bins
    bkg = [100.0] * n_bins
    bkgerr = [10.0] * n_bins
    sig = [30.0] * n_bins
    source = {
        'binning': binning,
        'bindata': {
            'data': data,
            'bkg': bkg,
            'bkgerr': bkgerr,
            'sig': sig
        }
    }

    pdf = hepdata_like(source['bindata']['sig'],
                       source['bindata']['bkg'],
                       source['bindata']['bkgerr'])
    data = source['bindata']['data'] + pdf.config.auxdata

    pyhf.runOnePoint(1.0, data, pdf,
                     pdf.config.suggested_init(),
                     pdf.config.suggested_bounds())

    # Reset backend
    pyhf.tensorlib = default_backend
```
**Traceback:**
```python-traceback
Traceback (most recent call last):
  File "/home/mcf/anaconda3/envs/pyhf/lib/python3.6/site-packages/scipy/optimize/slsqp.py", line 380, in _minimize_slsqp
    fx = float(np.asarray(fx))
TypeError: float() argument must be a string or a number, not 'Tensor'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "runOnePoint_example.py", line 36, in <module>
    pdf.config.suggested_bounds())
  File "/home/mcf/Code/GitHub/pyhf/pyhf/__init__.py", line 412, in runOnePoint
    pdf, init_pars, par_bounds))
  File "/home/mcf/Code/GitHub/pyhf/pyhf/__init__.py", line 380, in generate_asimov_data
    loglambdav, asimov_mu, data, pdf, init_pars, par_bounds)
  File "/home/mcf/Code/GitHub/pyhf/pyhf/optimize/opt_scipy.py", line 26, in constrained_bestfit
    method='SLSQP', args=(data, pdf), bounds=par_bounds)
  File "/home/mcf/anaconda3/envs/pyhf/lib/python3.6/site-packages/scipy/optimize/_minimize.py", line 495, in minimize
    constraints, callback=callback, **options)
  File "/home/mcf/anaconda3/envs/pyhf/lib/python3.6/site-packages/scipy/optimize/slsqp.py", line 382, in _minimize_slsqp
    raise ValueError("Objective function must return a scalar")
ValueError: Objective function must return a scalar
```
From the traceback it can be seen that [`opt_scipy`](https://github.com/diana-hep/pyhf/blob/master/pyhf/optimize/opt_scipy.py#L21) is getting used instead of [`opt_tflow`](https://github.com/diana-hep/pyhf/blob/master/pyhf/optimize/opt_tflow.py#L28).

# Checklist

- [x] Run `git fetch` to get the most up to date version of `master`
- [x] Searched through existing Issues to confirm this is not a duplicate issue
- [x] Filled out the Description, Expected Behavior, Actual Behavior, and Steps to Reproduce sections above or have edited/removed them in a way that fully describes the issue

**Comment by matthewfeickert** [comment]
From the traceback, it seems that the wrong optimizer is getting called. Apparently [`opt_scipy`](https://github.com/diana-hep/pyhf/blob/master/pyhf/optimize/opt_scipy.py#L21) is getting used instead of [`opt_tflow`](https://github.com/diana-hep/pyhf/blob/master/pyhf/optimize/opt_tflow.py#L28).

**Comment by matthewfeickert** [comment]
Closed by PR #100
